packages/core-plugin-api/src/translation/types.ts

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ResourceKey } from 'i18next';
+
+export type Locals = {
+  [language: string]: ResourceKey;
```

Comment:
I think what we want a potential `translationRef` to store are the message keys, so not the messages for each locale. For it to have a tangible benefit it would really need to have a type that contains all the messages embedded as well, so that you're able to reference that when defining your own messages.

Reply:
Cool, it's a great perspective. Maybe there are  two  scenarios

1.  plugins in Backstage repo or github which want to share to other developer. Just need to define message keys
2. plugins in internal company may  define message keys and values

Reply:
Yep but regardless the translation ref only needs to ember the available keys. The values would be brought in as part of locale definitions right?

Just to try to illustrate that with some code:

```ts
// Somewhere within todo plugin. This type is something that is likely NOT needed in practice, but instead inferred
// from the default messages. Depends a bit of whether we can provide docs that way or not though
export interface TodoPluginMessages {
  // Each message has its own docs that describe how it's used in the app
  /** Label used for individual todo items */
  todoItem: string
}
export const todoMessagesRef = createMessagesRef<TodoPluginMessages>({
  // no need to define an ID for this ref, it is injected when assigned to a plugin
  
  // Maybe default messages are provided here? In that case we can infer types here rather than explicit type.
  // This does cause duplication, but in practice we likely need to have the default set of messages available as fallback
  messages: {
    todoItem: 'TODO'
  },
})

// Plugin definition might look something like this?
export const todoPlugin = createPlugin({
  id: 'todo',
  messages: {
    ref: todoMessagesRef,
    locales: { // always lazy loaded
      zh: () => import('./translations/zh'),
    }
  }
})

// Then in the app we can likely provide overrides using the plugin message ref available as plugin.messages:
const app = createApp({
  messages: [{
    ref: todoPlugin.messages,
    messages: { // both sync and async loading supported
      todoItem: 'Todo'
    }
  }]
})
```

I ended up using "messages" as the term rather than "translations", just throwing it around as a potential option really, no super strong opinion on what to go with. The benefit of "messages" or something like it is that it's potentially more clear that it can be used for overrides in the same language too.

What do you think about that kind of structure and usage of translation refs?

Reply:
About the default message I think should be set in `t` function, like `<span>{ t('todoItem', 'TODO') }</span>`.  
Maybe there are two reasons
1. Default message in the context where it should be, may help developer to understand what the component's purpose
2. When I need to modify or delete the default message, it's more quick and reduce the probability of errors

Perhaps we do not need the `createMessagesRef `? code can be more simple to  plugin developers

```ts
// Somewhere within todo plugin. This type is something that is likely NOT needed in practice, but instead inferred
// from the default messages. Depends a bit of whether we can provide docs that way or not though
export interface TodoPluginMessages {
  // Each message has its own docs that describe how it's used in the app
  /** Label used for individual todo items */
  todoItem: string
}

// translations/zh.ts
export const messages:  TodoPluginMessages = {
   todoItem: 'TODO'
}

// Plugin definition might not need ref
export const todoPlugin = createPlugin({
  id: 'todo',
  locales: {
    zh: () => import('./translations/zh'),
  }
})

// Then in the app we can likely provide overrides using the plugin message ref available as plugin.messages:
const app = createApp({
  messages: [{
    ref: todoPlugin,
    messages: { // both sync and async loading supported
      todoItem: 'Todo'
    }
  }]
})
```
In `createApp` use `ref` to determine  the namespace of translation resource is a much better way. But one thing confused me, is message type necessary ? Missing message probably is tolerable ?

Reply:
Yep I agree it would be nicer to have the default message be defined where it's used. Let's see if we can work that in.

I do feel quite strongly still that we should try to build this with something like translation refs as part of the solution. The main idea there being that it provides a MUCH nicer experience for integrations that want to customize messages, as well as translators that write translations for plugins. Having all message keys be available in TypeScript both provides easy access to documentation, as well as ensures that you're not providing translations for unknown keys, and even that you're translating all keys. For example I think this could be a thing:

```ts
createApp({
  messages: [
    partialTranslation(todoPlugin, { // allows a subset of messages to be translated
      todoItem: 'Todo'
    }),
    fullTranslation(todoPlugin, { // does not compile unless all messages are translated
      todoItem: 'Todo'
    }),
  ]
})
```

Now the reason I think that translation refs are important for this kind of solution is to help ensure that the set of messages in TypeScript stay in sync with reality. If we do message key declarations in a separate type that is not directly tied to usage within the plugin, then they will inevitably fall out of sync. It would be far better if we can force the addition of a message key in the TypeScript world whenever a plugin builder wants to consume a new message in the component.

Played around with some executable code to illustrate:

```ts
interface TranslationRef<TKeys extends Record<string, {}>> {
  keys: TKeys
}

function createTranslationRef<TKeys extends Record<string, {}>>(): TranslationRef<TKeys> {
  return {keys: {} as any}
}

const todoTranslationsRef = createTranslationRef<{
  /** Shown as a label to represent todo */
  todoItem: {}
}>();

function createTranslations<TKeys extends Record<string, {}>>(ref: TranslationRef<TKeys>, messages: {[key in keyof TKeys]: string}): TKeys {
  return ref.keys;
}

createTranslations(todoTranslationsRef, {
  todoItem: 'Todo'
})
```

It's the experience of using `createTranslations` that I'm after here, where you both get the full set of keys, but also docs like this:

<img width="311" alt="image" src="https://github.com/backstage/backstage/assets/4984472/4844dfc8-71c7-4a73-a723-1a5987298784">

We of course also get a fairly neat usage of the refs in components as well, as you've already explored:

```ts
createTranslations(todoTranslationsRef, {
  todoItem: 'Todo'
})

function useTranslation<TKeys extends Record<string, {}>>(ref: TranslationRef<TKeys>): (key: keyof TKeys, defaultMessage: string) => string {
  return () => '';
}

const t = useTranslation(todoTranslationsRef);
```

<img width="461" alt="image" src="https://github.com/backstage/backstage/assets/4984472/48dd93df-dd0e-4a57-b06e-6782c7f07d09">

I do have to say that playing around with this I'm not actually sure that keeping the default message in the component is worth it. It does have the downside of not having the default available for lookup in the messages collection, but it also means you might be providing different defaults in different places, without realizing the mistake. When it comes to DX of components themselves and having the default messages available, consider the following:

```ts
const todoTranslationsRef = createTranslationRef<{
  /** Shown as a label to represent todo */
  todoItem: { default: 'Todo' }
}>();

function useTranslation<TKeys extends Record<string, {default: string}>>(ref: TranslationRef<TKeys>): <TKey extends keyof TKeys>(key: TKey) => TKeys[TKey]['default'] {
  return () => '' as any;
}

const t = useTranslation(todoTranslationsRef);

t('todoItem')
```

Here you actually get fairly good visibility into the default message, although it does require you to be within an IDE:

<img width="384" alt="image" src="https://github.com/backstage/backstage/assets/4984472/b45efc74-7105-4d56-bb97-d79e3c778ef9">

Reply:
We do not consider to use `http-backend` plugin currently, due to json file has less type tips. 

【Having all message keys be available in TypeScript both provides easy access to documentation】It‘s indeed make sense. We need to do something like that defined a function or other approach app developer who want to customize messages must have to use and in the function context can get plugin's translation keys to check which key is invalid or valid.

Restrict the app developer custom way but make sure that he do the right way.

Reply:
> Restrict the app developer custom way but make sure that he do the right way.

I'm not entirely sure exactly what kind of restriction you have in mind there, but otherwise it sounds like we're on the same page.

Basically as with a lot of other parts of Backstage we optimize for the best app integration and customization experience. This will mean it's a bit more tricky to add i18n to plugins, but it will be a lot smoother to customize messages and add translations.

Reply:
Just recovery from the COVID-19. I will apply our new opinions and update this PR

Reply:
<img width="780" alt="image" src="https://github.com/backstage/backstage/assets/7720722/5a5c1fff-7037-48b0-abb1-71812a6f7b90">
<img width="553" alt="image" src="https://github.com/backstage/backstage/assets/7720722/61ec8ab3-0f03-4d2a-b573-fdeff587bad8">

In `createApp` function, backstage adopter can get tips about message keys which defined in plugin.

<img width="602" alt="image" src="https://github.com/backstage/backstage/assets/7720722/4099b6f5-b114-4408-8b0f-1f5e778ad2db">

Reply:
As mentioned I played around with this some more, and I'm arriving at this probably being the best design in the end:

```ts
function createTranslationRef<
  const TMessages extends Record<string, string>,
>(options: { messages: TMessages }): TranslationRef<TMessages> {
  return options;
}

const todoTranslationsRef = createTranslationRef({
  messages: {
    /** Shown as a label to represent todo */
    todoItem: 'TODO',
  },
});
```

Note the `const` keyword in the type parameter of `createTranslationRef`, it removes the need for an explicit `as const` when defining the messages. Unfortunately this is only available in TypeScript 5.0, and with our upcoming policy update we'll need to support 4.9 as well. I don't think this is a huge issue though, users will simply have to use `as const` until we can remove it.

Why I think this design in good:

It provides a single definition of type and runtime values by avoiding a separate declaration of the default messages. The `TranslationRef`s contain all the information that I think we want, both all available messages keys, as well as TSDoc comments for those keys, and the default message. This information both allows us to provide a good experience when defining messages, as we've already explored, where you receive hints for the available message keys **as well as their documentation**. By adding the default messages to the translation ref we're also able to receive hints during implementation so that you're at least able to see the default messages as type hints inside components.

On the consuming end I'm still seeing a few options I'll just highlight the two main ones:

### String-based

This is the more traditional `useTranslation` as we've seen it before:

```ts
function useTranslation<TMessages extends Record<string, string>>(
  ref: TranslationRef<TMessages>,
): <TKey extends keyof TMessages>(key: TKey) => TMessages[TKey] {
  return () => '' as any; // TODO
}

function MyComponent() {
  const t = useTranslation(todoTranslationsRef);

  return <div>{t('todoItem')}</div>;
}
```

With this approach we'll receive a hint what the default message is, as mentioned above:

<img width="482" alt="image" src="https://github.com/backstage/backstage/assets/4984472/3f6597f1-ddd2-4681-b6de-681e9043836e">

However, as you can see we will not receive a documentation hint.

### Property-based

This is of course not the typical i18n API, since that's set up to be more dynamic, but with our translation refs we're able to do this:

```ts
function useTranslations<TMessages extends Record<string, string>>(
  ref: TranslationRef<TMessages>,
): {[key in keyof TMessages]: () => TMessages[key]} {
  return {} as any; // TODO
}

function MyComponent() {
  const t = useTranslations(todoTranslationsRef);

  return <div>{t.todoItem()}</div>;
}
```

With this we'll both see the default message, but also the documentation for each of the messages:

<img width="393" alt="image" src="https://github.com/backstage/backstage/assets/4984472/dfc10efd-5ed1-40c2-bd88-f778150e57d5">

The reason for it being methods rather than constant properties is to support options and templating values. Which we by the way might be able to get a bit creative with given TypeScript's string templating type powers 😁 

<br>
<br>

<!--
interface TranslationRef<TMessages extends Record<string, string>> {
  messages: TMessages;
}

function createTranslationRef<
  const TMessages extends Record<string, string>,
>(options: { messages: TMessages }): TranslationRef<TMessages> {
  return options;
}

const todoTranslationsRef = createTranslationRef({
  messages: {
    /** Shown as a label to represent todo */
    todoItem: 'TODO',

    /** Derp */
    otherMessage: 'DERP'
  },
});

function createTranslations<TMessages extends Record<string, string>>(
  ref: TranslationRef<TMessages>,
  messages: { [key in keyof TMessages]: string },
): TMessages {
  return ref.messages;
}

createTranslations(todoTranslationsRef, {
  todoItem: 'Todo',
});

function useTranslation<TMessages extends Record<string, string>>(
  ref: TranslationRef<TMessages>,
): <TKey extends keyof TMessages>(key: TKey) => TMessages[TKey] {
  return () => '' as any; // TODO
}

function MyComponent2() {
  const t = useTranslation(todoTranslationsRef);

  return <div>{t('todoItem')}</div>;
}


function useTranslations<TMessages extends Record<string, string>>(
  ref: TranslationRef<TMessages>,
): {[key in keyof TMessages]: () => TMessages[key]} {
  return {} as any; // TODO
}

function MyComponent() {
  const t = useTranslations(todoTranslationsRef);

  return <div>{t.todoItem()}</div>;
}
-->

So... curious to hear about what you think about this approach to `createTranslationRef` and which consumption approach is your favorite.

Reply:
About `createTranslationRef `, if plugin developer want to support the messages inherently in the plugin, for example supporting en and de messages. How does `createTranslationRef ` options look like ?

About consuming , my favorite consumption approach is `string-based`. Depend on mostly react developer‘s habit. What I think is that `string-based` maybe more acceptable for backstage plugin developers and adopters